lib/grizzly/events.ex

```diff
@@ -17,40 +17,174 @@ defmodule Grizzly.Events do
+      {type, Enum.sort(keys)}
+    end
+  end
+
+  @doc """
+  Subscribe to an event, all commands from a node, or a specific command from any
+  node.
+
+  ## Options
+
+  * `:firehose` - Normally, subscribers will only receive unsolicited commands
+    from the node (or command) they are subscribed to. If this option is set to
+    true, subscribers will receive all incoming commands that match, including
+    responses to commands sent via `Grizzly.send_command/4`. Defaults to `false`.
+  """
+  @spec subscribe(subject() | [subject()], subscribe_opts()) :: :ok
+  def subscribe(subject, opts \\ [])
+
+  def subscribe([subject | subjects], opts) do
+    subscribe(subject, opts)
+    subscribe(subjects, opts)
+  end
+
+  def subscribe([], _opts), do: :ok
+
+  def subscribe(node_id, opts) when is_node_id(node_id) do
+    subscribe({:node, node_id}, opts)
+  end
+
+  def subscribe(event, _opts) when is_event(event) do
+    _ = Registry.register(__MODULE__, event, [])
+    :ok
+  end
+
+  def subscribe(subject, opts) do
+    _ = Registry.register(__MODULE__, subject, Map.new(opts))
+    :ok
   end
 
   @doc """
   Unsubscribe from one or more Grizzly events.
   """
-  @spec unsubscribe(event() | [event()]) :: :ok
-  def unsubscribe(event_or_events) do
-    event_or_events
-    |> List.wrap()
-    |> Enum.each(&Registry.unregister(__MODULE__, &1))
+  @spec unsubscribe(subject() | [subject()]) :: :ok
+  def unsubscribe(subject)
+
+  def unsubscribe(node_id) when is_node_id(node_id),
+    do: unsubscribe({:node, node_id})
+
+  def unsubscribe(subjects) when is_list(subjects),
+    do: Enum.each(subjects, &unsubscribe/1)
+
+  def unsubscribe(subject) do
+    _ = Registry.unregister(__MODULE__, subject)
+    :ok
   end
 
   @doc false
-  @spec broadcast(event(), term()) :: :ok
-  def broadcast(event, payload) do
+  @spec broadcast_event(event(), term()) :: :ok
+  def broadcast_event(event, payload) when is_event(event) do
     Registry.dispatch(__MODULE__, event, fn entries ->
       for {pid, _} <- entries do
         send(pid, {:grizzly, event, payload})
       end
     end)
   end
+
+  @doc false
+  @spec broadcast_report(Report.t()) :: :ok
+  def broadcast_report(%Report{command: %Command{}} = report) do
+    report
+    |> __subs_for_report__()
+    |> Enum.each(&send(&1, {:grizzly, :report, report}))
+  end
+
+  def broadcast_report(_), do: :ok
+
+  @doc false
+  @spec __subs_for_report__(Report.t()) :: [pid()]
+  def __subs_for_report__(
+        %Report{node_id: node_id, command: %Command{name: command_name}} = report
+      ) do
+    guards =
+      if report.type == :unsolicited do
+        match_spec_guards_for_unsolicited(node_id, command_name)
+      else
+        match_spec_guards_for_firehose(node_id, command_name)
+      end
+
+    subscribers =
+      Registry.select(Grizzly.Events, [
+        {
+          {:"$1", :"$2", :"$3"},
+          guards,
+          [{{:"$1", :"$2", :"$3"}}]
+        }
+      ])
```

Comment:
Match specs aren't too bad to grok if I stare long enough, but I'm curious what you think about options like [`:ets.fun2ms/1`](https://www.erlang.org/doc/apps/stdlib/ets.html#fun2ms/1) or [`:ex2ms`](https://github.com/ericmj/ex2ms)?